// @tomorrowos/sdk/dist/store/factory.js
import path from "path";
import { MemoryStore } from "./memory-store.js";
import { PostgresStore } from "./postgres-store.js";
import { SQLiteStore } from "./sqlite-store.js";
export function defaultSQLitePath(cwd = process.cwd()) {
    return path.join(cwd, "data", "tomorrowos.db");
}
export function createTomorrowOSStore(options = {}) {
    const env = options.env ?? process.env;
    const databaseUrl = options.databaseUrl ?? env.SUPABASE_URL ?? env.DATABASE_URL;
    const driver = (options.driver ??
        env.TOMORROWOS_STORE ??
        (databaseUrl ? "postgres" : "sqlite")).toLowerCase();
    if (driver === "memory") {
        return new MemoryStore();
    }
    if (driver === "postgres" || driver === "supabase") {
        if (!databaseUrl) {
            throw new Error(`TOMORROWOS_STORE=${driver} requires SUPABASE_URL or DATABASE_URL. On Replit prefer SUPABASE_URL (DATABASE_URL is often reserved). Or unset TOMORROWOS_STORE to use SQLite.`);
        }
        return new PostgresStore({
            connectionString: databaseUrl,
            ssl: options.postgresSsl ?? resolvePostgresSsl(env, databaseUrl)
        });
    }
    if (driver === "sqlite") {
        return new SQLiteStore({
            databasePath: options.sqlitePath ?? env.TOMORROWOS_DB_PATH ?? defaultSQLitePath()
        });
    }
    throw new Error(`Unsupported TOMORROWOS_STORE "${driver}". Use "sqlite", "supabase", "postgres", "memory", or pass a custom TomorrowOSStore to new TomorrowOS({ store }).`);
}
function resolvePostgresSsl(env, databaseUrl) {
    const raw = env.DATABASE_SSL?.trim().toLowerCase();
    if (raw === "false" || raw === "0" || raw === "off")
        return false;
    if (raw === "true" || raw === "1" || raw === "on") {
        return { rejectUnauthorized: false };
    }
    if (/supabase\.(co|com)|sslmode=require/i.test(databaseUrl)) {
        return { rejectUnauthorized: false };
    }
    return undefined;
}
