// @tomorrowos/sdk/dist/store/memory-store.js
/**
 * Default store for development / single-node demos.
 * Data is lost on process restart — not for multi-instance production.
 */
export class MemoryStore {
    pendingCodes = new Map();
    deviceRegistry = new Map();
    codeToDeviceId = new Map();
    pairedDevices = new Map();
    playlists = new Map();
    uploadedAssets = new Map();
    deviceAssignments = new Map();
    async setPendingCode(code, record) {
        this.pendingCodes.set(code, record);
    }
    async getPendingCode(code) {
        return this.pendingCodes.get(code);
    }
    async deletePendingCode(code) {
        this.pendingCodes.delete(code);
    }
    async getDeviceRegistry(deviceId) {
        return this.deviceRegistry.get(deviceId);
    }
    async setDeviceRegistry(deviceId, record) {
        const existing = this.deviceRegistry.get(deviceId);
        if (existing?.permanentPairingCode) {
            this.codeToDeviceId.delete(existing.permanentPairingCode);
        }
        this.deviceRegistry.set(deviceId, record);
        this.codeToDeviceId.set(record.permanentPairingCode, deviceId);
    }
    async getDeviceRegistryByCode(code) {
        const deviceId = this.codeToDeviceId.get(code);
        if (!deviceId)
            return undefined;
        const record = this.deviceRegistry.get(deviceId);
        if (!record)
            return undefined;
        return { deviceId, record };
    }
    async setPairedDevice(deviceId, record) {
        this.pairedDevices.set(deviceId, record);
    }
    async getPairedDevice(deviceId) {
        return this.pairedDevices.get(deviceId);
    }
    async deletePairedDevice(deviceId) {
        this.pairedDevices.delete(deviceId);
    }
    async listPairedDevices() {
        return [...this.pairedDevices.entries()].map(([deviceId, record]) => ({
            deviceId,
            record
        }));
    }
    async listPlaylists() {
        return [...this.playlists.values()];
    }
    async getPlaylist(id) {
        return this.playlists.get(id);
    }
    async setPlaylist(record) {
        this.playlists.set(record.id, record);
    }
    async isPlaylistNameTaken(name, excludeId) {
        const target = name.trim().toLowerCase();
        for (const playlist of this.playlists.values()) {
            if (playlist.retired)
                continue;
            if (excludeId && playlist.id === excludeId)
                continue;
            if (playlist.name.trim().toLowerCase() === target)
                return true;
        }
        return false;
    }
    async getUploadedAsset(id) {
        return this.uploadedAssets.get(id);
    }
    async getUploadedAssetBySha256(sha256) {
        for (const asset of this.uploadedAssets.values()) {
            if (asset.sha256 === sha256) {
                return asset;
            }
        }
        return undefined;
    }
    async setUploadedAsset(record) {
        this.uploadedAssets.set(record.id, record);
    }
    async deleteUploadedAsset(id) {
        this.uploadedAssets.delete(id);
    }
    async getDeviceAssignments(deviceId) {
        return [...(this.deviceAssignments.get(deviceId) ?? [])];
    }
    async setDeviceAssignments(deviceId, assignments) {
        this.deviceAssignments.set(deviceId, assignments.map((a) => ({
            ...a,
            snapshot: {
                ...a.snapshot,
                items: a.snapshot.items.map((item) => ({ ...item }))
            }
        })));
    }
}
