// @tomorrowos/sdk/dist/playlist-catalog.js
import { randomUUID } from "crypto";
function normalizeMediaBaseUrl(raw) {
    let s = String(raw || "").trim();
    if (!s)
        return "";
    if (!/^https?:\/\//i.test(s))
        s = `http://${s}`;
    try {
        return new URL(s).origin;
    }
    catch {
        return "";
    }
}
function absolutizeMediaUrl(url, mediaBaseUrl) {
    const trimmed = String(url || "").trim();
    if (!trimmed)
        return trimmed;
    if (/^https?:\/\//i.test(trimmed))
        return trimmed;
    const base = normalizeMediaBaseUrl(mediaBaseUrl);
    if (!base)
        return trimmed;
    return `${base}${trimmed.startsWith("/") ? trimmed : `/${trimmed}`}`;
}
function absolutizePlaylistItems(items, mediaBaseUrl) {
    const base = normalizeMediaBaseUrl(mediaBaseUrl);
    if (!base)
        return items.map((item) => ({ ...item }));
    return items.map((item) => ({
        ...item,
        url: absolutizeMediaUrl(item.url, base)
    }));
}
function cloneSnapshot(playlist, mediaBaseUrl) {
    return {
        id: playlist.id,
        name: playlist.name,
        schedule: playlist.schedule ? { ...playlist.schedule } : undefined,
        items: absolutizePlaylistItems(playlist.items, mediaBaseUrl)
    };
}
/** True when the playlist has no start/end date or time (always-on / default content). */
function isUnscheduledPlaylist(schedule) {
    if (!schedule)
        return true;
    return !(schedule.startDate ||
        schedule.endDate ||
        schedule.start ||
        schedule.end);
}
export class PlaylistCatalog {
    store;
    constructor(store) {
        this.store = store;
    }
    async listPlaylists() {
        const all = await this.store.listPlaylists();
        return all.filter((p) => !p.retired);
    }
    async listPlaylistsIncludingRetired() {
        return this.store.listPlaylists();
    }
    async getPlaylist(id) {
        return this.store.getPlaylist(id);
    }
    async savePlaylist(input) {
        const name = String(input.name || "").trim();
        if (!name) {
            throw Object.assign(new Error("Playlist name is required"), {
                code: "PLAYLIST_INVALID"
            });
        }
        const id = input.id?.trim() || randomUUID();
        const taken = await this.store.isPlaylistNameTaken(name, id);
        if (taken) {
            throw Object.assign(new Error("Playlist name already exists"), {
                code: "PLAYLIST_NAME_CONFLICT"
            });
        }
        const now = new Date().toISOString();
        const record = {
            id,
            name,
            schedule: input.schedule,
            items: Array.isArray(input.items) ? input.items : [],
            updatedAt: now,
            retired: false
        };
        await this.store.setPlaylist(record);
        return record;
    }
    async retirePlaylist(id) {
        const existing = await this.store.getPlaylist(id);
        if (!existing) {
            throw Object.assign(new Error("Playlist not found"), { code: "PLAYLIST_NOT_FOUND" });
        }
        const record = {
            ...existing,
            retired: true,
            retiredAt: new Date().toISOString()
        };
        await this.store.setPlaylist(record);
        return record;
    }
    async getDeviceAssignments(deviceId) {
        return this.store.getDeviceAssignments(deviceId);
    }
    async publishPlaylistsToDevice(deviceId, playlistIds, options = {}) {
        const incoming = [...new Set(playlistIds.map((x) => String(x).trim()).filter(Boolean))];
        if (incoming.length === 0) {
            throw Object.assign(new Error("Select at least one playlist"), {
                code: "PLAYLIST_INVALID"
            });
        }
        const existing = await this.store.getDeviceAssignments(deviceId);
        const existingById = new Map(existing.map((a) => [a.playlistId, a]));
        // Resolve incoming playlists first so we know whether this publish includes
        // an always-on (no start/end) playlist that should override prior always-on ones.
        const incomingSnapshots = new Map();
        for (const playlistId of incoming) {
            const playlist = await this.store.getPlaylist(playlistId);
            if (!playlist || playlist.retired) {
                throw Object.assign(new Error(`Playlist not available: ${playlistId}`), {
                    code: "PLAYLIST_NOT_FOUND"
                });
            }
            incomingSnapshots.set(playlistId, cloneSnapshot(playlist, options.mediaBaseUrl));
        }
        const incomingHasUnscheduled = [...incomingSnapshots.values()].some((snap) => isUnscheduledPlaylist(snap.schedule));
        const allIds = [...new Set([...existing.map((a) => a.playlistId), ...incoming])];
        const publishedAt = new Date().toISOString();
        const assignments = [];
        for (const playlistId of allIds) {
            if (!incoming.includes(playlistId)) {
                const kept = existingById.get(playlistId);
                if (!kept)
                    continue;
                // Later publish of an unscheduled playlist replaces earlier unscheduled ones.
                // Scheduled (dated) assignments are preserved.
                if (incomingHasUnscheduled &&
                    isUnscheduledPlaylist(kept.snapshot.schedule)) {
                    continue;
                }
                assignments.push(kept);
                continue;
            }
            assignments.push({
                playlistId,
                publishedAt,
                snapshot: incomingSnapshots.get(playlistId)
            });
        }
        await this.store.setDeviceAssignments(deviceId, assignments);
        return this.buildPolicyFromAssignments(assignments, {
            useLatest: false,
            mediaBaseUrl: options.mediaBaseUrl
        });
    }
    async removePlaylistFromDevice(deviceId, playlistId) {
        const assignments = await this.store.getDeviceAssignments(deviceId);
        const next = assignments.filter((a) => a.playlistId !== playlistId);
        await this.store.setDeviceAssignments(deviceId, next);
        return this.buildPolicyFromAssignments(next, { useLatest: false });
    }
    /** Remove every playlist assignment from a device (CMS catalog + empty policy). */
    async clearAllAssignmentsFromDevice(deviceId) {
        await this.store.setDeviceAssignments(deviceId, []);
        return this.buildPolicyFromAssignments([], { useLatest: false });
    }
    /**
     * If the device has this playlist assigned, refresh its snapshot to latest
     * and return a rebuilt snapshot policy for immediate push.
     */
    async refreshPlaylistSnapshotOnDevice(deviceId, playlistId, options = {}) {
        const assignments = await this.store.getDeviceAssignments(deviceId);
        const idx = assignments.findIndex((a) => a.playlistId === playlistId);
        if (idx < 0)
            return { updated: false };
        const playlist = await this.store.getPlaylist(playlistId);
        if (!playlist || playlist.retired) {
            return { updated: false };
        }
        const next = assignments.map((a, i) => i !== idx
            ? a
            : {
                playlistId,
                publishedAt: new Date().toISOString(),
                snapshot: cloneSnapshot(playlist, options.mediaBaseUrl)
            });
        await this.store.setDeviceAssignments(deviceId, next);
        const built = await this.buildPolicyFromAssignments(next, {
            useLatest: false,
            mediaBaseUrl: options.mediaBaseUrl
        });
        return { updated: true, built };
    }
    async buildPolicyForDevice(deviceId, options = {}) {
        const assignments = await this.store.getDeviceAssignments(deviceId);
        return this.buildPolicyFromAssignments(assignments, options);
    }
    async buildPolicyFromAssignments(assignments, options) {
        const useLatest = options.useLatest === true;
        const mediaBaseUrl = options.mediaBaseUrl;
        const playlists = [];
        for (const assignment of assignments) {
            if (useLatest) {
                const current = await this.store.getPlaylist(assignment.playlistId);
                if (current && !current.retired) {
                    playlists.push(cloneSnapshot(current, mediaBaseUrl));
                    continue;
                }
            }
            playlists.push({
                ...assignment.snapshot,
                items: absolutizePlaylistItems(assignment.snapshot.items, mediaBaseUrl)
            });
        }
        return {
            policy: {
                playlists,
                fallback: { type: "brand" },
                revision: Date.now(),
                syncMode: useLatest ? "latest" : "snapshot"
            }
        };
    }
    canPublishPlaylistToNewDevice(playlistId) {
        return this.store.getPlaylist(playlistId).then((p) => !!p && !p?.retired);
    }
}
